### DISCUSSION TITLE
Couldn't capture workqueue/* ftrace event in perfetto-trace with userroot build.

### DISCUSSION THREAD

**Discussion by fx9037curry** [google/perfetto#953] [Q&A] [answered]
Using command below:
adb shell "perfetto -o /data/misc/perfetto-traces/trace.perfetto-trace -b 512mb -t 10s sched freq idle am wm gfx view binder_driver hal input **workqueue/*** "

In eng build, I can capture all of the Ftrace Events in perfetto-trace.
But in userroot build, the "workqueue/*" ftrace events are missing.
Pls help me to solve this issue, thx a lot.

**Comment by fx9037curry** [comment]
In eng version:
![Image](https://github.com/user-attachments/assets/022d56bc-6fe6-4ce8-af29-b3b28be5539f)

In userroot version with same cmdline:
![Image](https://github.com/user-attachments/assets/dd63de1f-e13b-4aba-93d7-5d01672c0eb6)

**Comment by LalitMaganti** [comment, marked as answer]
Due to security reasons, on user builds, Perfetto on the system only has support for enabling a limited set of ftrace events it is explicitly allowed to read in init.rc files.

This list of events is controlled via UNIX ACLs and live in the atrace.rc [1] file in the Android build. Presumably you have control over the the way the system is built given you are able to produce engineering builds so you should add the events you want to this list.

Alternatively, you could also collect these events with a sideloaded tracebox binary but probably you want to do the above, I'm just mentioning it for completeness sake.

[1] https://cs.android.com/android/platform/superproject/main/+/main:frameworks/native/cmds/atrace/atrace.rc;l=28?q=f:atrace.rc

**Reply by fx9037curry** [reply]
thanks a lot, i'll have a try

**Reply by Devister** [reply]
Hi, I have added my ftrace events in atrace.rc, Perfetto can read these events on userroot builds, but still failed on user builds.
Please help me to solve this issue, thx a lot.

**Reply by LalitMaganti** [reply]
Forgot to mention but you should also enable them in sepolicy as debugfs_tracing (see examples at https://cs.android.com/android/platform/superproject/main/+/main:system/sepolicy/private/genfs_contexts;l=235?q=system%2Fsepolicy%2Fprivate%2Fgenfs_contexts)

**Reply by Devister** [reply]
thank a lot, I'll  have a try
